spring-cloud-stream-binder-rabbit-core/src/main/java/org/springframework/cloud/stream/binder/rabbit/provisioning/RabbitExchangeQueueProvisioner.java

```diff
@@ -0,0 +1,514 @@
+		return prefix + name;
+	}
+
+	private Exchange buildExchange(RabbitCommonProperties properties, String exchangeName) {
+		try {
+			ExchangeBuilder builder = new ExchangeBuilder(exchangeName, properties.getExchangeType());
+			if (properties.isDelayedExchange()) {
+				builder.delayed();
+			}
+			return builder.build();
+		}
+		catch (Exception e) {
+			throw new ProvisioningException("Failed to create exchange object", e);
+		}
+	}
+
+	private void declareExchange(final String rootName, final Exchange exchange) {
+		try {
+			this.rabbitAdmin.declareExchange(exchange);
+		}
+		catch (AmqpConnectException e) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug(
+						"Declaration of exchange: " + exchange.getName() + " deferred - connection not available");
+			}
+		}
+		addToAutoDeclareContext(rootName + ".exchange", exchange);
+	}
+
+	private void addToAutoDeclareContext(String name, Object bean) {
+		synchronized (this.autoDeclareContext) {
+			if (!this.autoDeclareContext.containsBean(name)) {
+				this.autoDeclareContext.getBeanFactory().registerSingleton(name, bean);
+			}
+		}
+	}
+
+	private void declareBinding(String rootName, org.springframework.amqp.core.Binding binding) {
+		try {
+			this.rabbitAdmin.declareBinding(binding);
+		}
+		catch (AmqpConnectException e) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug(
+						"Declaration of binding: " + rootName + ".binding deferred - connection not available");
+			}
+		}
+		addToAutoDeclareContext(rootName + ".binding", binding);
+	}
+
+	public void cleanAutoDeclareContext(String name) {
+		synchronized (this.autoDeclareContext) {
+			removeSingleton(name + ".binding");
+			removeSingleton(name);
+			String dlq = name + ".dlq";
+			removeSingleton(dlq + ".binding");
+			removeSingleton(dlq);
+		}
+	}
+
+	private void removeSingleton(String name) {
+		if (this.autoDeclareContext.containsBean(name)) {
+			ConfigurableListableBeanFactory beanFactory = this.autoDeclareContext.getBeanFactory();
+			if (beanFactory instanceof DefaultListableBeanFactory) {
+				((DefaultListableBeanFactory) beanFactory).destroySingleton(name);
+			}
+		}
+	}
+
+	private final class RabbitProducerDestination implements ProducerDestination {
+
+		private final Exchange exchange;
+		private final Binding binding;
+
+		private RabbitProducerDestination(Exchange exchange, Binding binding) {
+			Assert.notNull(exchange, "exchange must not be null");
+			this.exchange = exchange;
+			this.binding = binding;
+		}
+
+		@Override
+		public String getName() {
+			return this.exchange.getName();
+		}
+
+		@Override
+		public String getNameForPartition(int partition) {
+			return this.exchange.getName();
+		}
+
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder();
+			sb.append(this.exchange.toString());
+			if (this.binding != null) {
+				sb.append(this.binding.toString());
+			}
+			return sb.toString();
```

Comment:
I think that `toString()` should include the info about root class as well:
```
public String toString() {
			StringBuilder sb = new StringBuilder("RabbitProducerDestination{");
			sb.append("exchange=").append(this.exchange);
			if (this.binding != null) {
				sb.append(", binding=").append(this.binding);
			}
			sb.append('}');
			return sb.toString();
		}
```
Otherwise it fully unclear what is that.
Plus you are missing `,` between props.